File: examples/canonical-auth/python/src/routes/groups.py
from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from typing import Optional
import uuid
from datetime import datetime, timezone

from src.auth.claims import AuthContext
from src.auth.middleware import require_permission, store, get_auth_context
from src.domain.models import Group, GroupPermission, UserGroup

router = APIRouter()

class CreateGroupRequest(BaseModel):
    name: str
    slug: str
    permissions: Optional[list[str]] = None

class AssignPermissionRequest(BaseModel):
    permission_id: str

class AssignUserRequest(BaseModel):
    user_id: str

@router.get("/groups", dependencies=[Depends(require_permission("iam:group:read"))])
def list_groups(auth: AuthContext = Depends(get_auth_context)):
    tenant_groups = [g for g in store.groups if g.tenant_id == auth.tenant_id]
    return {"groups": tenant_groups}

@router.post("/groups", dependencies=[Depends(require_permission("iam:group:create"))])
def create_group(request: CreateGroupRequest, auth: AuthContext = Depends(get_auth_context)):
    now = datetime.now(timezone.utc).isoformat()
    
    if any(g.tenant_id == auth.tenant_id and g.slug == request.slug for g in store.groups):
        raise HTTPException(status_code=400, detail="Group slug already exists in tenant")
        
    group_id = str(uuid.uuid4())
    new_group = Group(
        id=group_id,
        tenant_id=auth.tenant_id,
        slug=request.slug,
        name=request.name,
        created_at=now
    )
    store.groups.append(new_group)
    
    if request.permissions:
        for perm_name in request.permissions:
            perm = next((p for p in store.permissions if p.name == perm_name), None)
            if not perm:
                raise HTTPException(status_code=400, detail=f"Permission {perm_name} not found")
            store.group_permissions.append(GroupPermission(
                id=str(uuid.uuid4()),
                group_id=group_id,
                permission_id=perm.id,
                granted_at=now
            ))
            
    return new_group

@router.post("/groups/{group_id}/permissions", dependencies=[Depends(require_permission("iam:group:assign_permission"))])
def assign_permission(group_id: str, request: AssignPermissionRequest, auth: AuthContext = Depends(get_auth_context)):
    group = next((g for g in store.groups if g.id == group_id and g.tenant_id == auth.tenant_id), None)
    if not group:
         raise HTTPException(status_code=404, detail="Group not found in your tenant")
         
    perm = next((p for p in store.permissions if p.id == request.permission_id), None)
    if not perm:
         raise HTTPException(status_code=400, detail="Permission not found")
         
    if any(gp.group_id == group_id and gp.permission_id == perm.id for gp in store.group_permissions):
        return {"status": "ok", "message": "Already assigned"}

    now = datetime.now(timezone.utc).isoformat()
    store.group_permissions.append(GroupPermission(
        id=str(uuid.uuid4()),
        group_id=group_id,
        permission_id=perm.id,
        granted_at=now
    ))
    
    # Increment acl_ver for users in this group
    for ug in store.user_groups:
        if ug.group_id == group_id:
            user = store.get_user_by_id(ug.user_id)
            if user:
                user.acl_ver += 1
                
    return {"status": "ok"}

@router.post("/groups/{group_id}/users", dependencies=[Depends(require_permission("iam:group:assign_user"))])
def assign_user(group_id: str, request: AssignUserRequest, auth: AuthContext = Depends(get_auth_context)):
    group = next((g for g in store.groups if g.id == group_id and g.tenant_id == auth.tenant_id), None)
    if not group:
         raise HTTPException(status_code=404, detail="Group not found in your tenant")
         
    if not store.verify_membership(request.user_id, auth.tenant_id):
         raise HTTPException(status_code=404, detail="User not found in your tenant")
         
    if any(ug.group_id == group_id and ug.user_id == request.user_id for ug in store.user_groups):
        return {"status": "ok", "message": "Already assigned"}
        
    now = datetime.now(timezone.utc).isoformat()
    store.user_groups.append(UserGroup(
        id=str(uuid.uuid4()),
        user_id=request.user_id,
        group_id=group_id,
        joined_at=now
    ))
    
    user = store.get_user_by_id(request.user_id)
    if user:
        user.acl_ver += 1
        
    return {"status": "ok"}

File: examples/canonical-auth/typescript/src/routes/groups.ts
import { Hono } from "hono";
import { jwtMiddleware, requirePermission } from "../auth/middleware.ts";
import { store } from "../domain/store.ts";
import type { AuthContext } from "../auth/context.ts";
import type { Group, GroupPermission, UserGroup } from "../domain/models.ts";

const app = new Hono();

app.use("/groups", jwtMiddleware);
app.use("/groups/:id/permissions", jwtMiddleware);
app.use("/groups/:id/users", jwtMiddleware);

app.get("/groups", requirePermission("iam:group:read"), async (c) => {
  const auth = c.get("auth") as AuthContext;
  const tenantGroups = store.groups.filter((g) => g.tenant_id === auth.tenantId);
  return c.json({ groups: tenantGroups });
});

app.post("/groups", requirePermission("iam:group:create"), async (c) => {
  const auth = c.get("auth") as AuthContext;
  const body = await c.req.json();
  const now = new Date().toISOString().replace(".000Z", "Z");

  if (store.groups.some((g) => g.tenant_id === auth.tenantId && g.slug === body.slug)) {
    return c.json({ detail: "Group slug already exists in tenant" }, 400);
  }

  const groupId = crypto.randomUUID();
  const newGroup: Group = {
    id: groupId,
    tenant_id: auth.tenantId as string,
    slug: body.slug,
    name: body.name,
    created_at: now,
  };
  store.groups.push(newGroup);

  if (body.permissions && Array.isArray(body.permissions)) {
    for (const permName of body.permissions) {
      const perm = store.permissions.find((p) => p.name === permName);
      if (!perm) {
        return c.json({ detail: `Permission ${permName} not found` }, 400);
      }
      store.group_permissions.push({
        id: crypto.randomUUID(),
        group_id: groupId,
        permission_id: perm.id,
        granted_at: now,
      });
    }
  }

  return c.json(newGroup);
});

app.post("/groups/:id/permissions", requirePermission("iam:group:assign_permission"), async (c) => {
  const auth = c.get("auth") as AuthContext;
  const groupId = c.req.param("id");
  const body = await c.req.json();

  const group = store.groups.find((g) => g.id === groupId && g.tenant_id === auth.tenantId);
  if (!group) {
    return c.json({ detail: "Group not found in your tenant" }, 404);
  }

  const perm = store.permissions.find((p) => p.id === body.permission_id);
  if (!perm) {
    return c.json({ detail: "Permission not found" }, 400);
  }

  if (store.group_permissions.some((gp) => gp.group_id === groupId && gp.permission_id === perm.id)) {
    return c.json({ status: "ok", message: "Already assigned" });
  }

  const now = new Date().toISOString().replace(".000Z", "Z");
  store.group_permissions.push({
    id: crypto.randomUUID(),
    group_id: groupId,
    permission_id: perm.id,
    granted_at: now,
  });

  for (const ug of store.user_groups) {
    if (ug.group_id === groupId) {
      const user = store.getUserById(ug.user_id);
      if (user) {
        user.acl_ver += 1;
      }
    }
  }

  return c.json({ status: "ok" });
});

app.post("/groups/:id/users", requirePermission("iam:group:assign_user"), async (c) => {
  const auth = c.get("auth") as AuthContext;
  const groupId = c.req.param("id");
  const body = await c.req.json();

  const group = store.groups.find((g) => g.id === groupId && g.tenant_id === auth.tenantId);
  if (!group) {
    return c.json({ detail: "Group not found in your tenant" }, 404);
  }

  if (!store.verifyMembership(body.user_id, auth.tenantId as string)) {
    return c.json({ detail: "User not found in your tenant" }, 404);
  }

  if (store.user_groups.some((ug) => ug.group_id === groupId && ug.user_id === body.user_id)) {
    return c.json({ status: "ok", message: "Already assigned" });
  }

  const now = new Date().toISOString().replace(".000Z", "Z");
  store.user_groups.push({
    id: crypto.randomUUID(),
    user_id: body.user_id,
    group_id: groupId,
    joined_at: now,
  });

  const user = store.getUserById(body.user_id);
  if (user) {
    user.acl_ver += 1;
  }

  return c.json({ status: "ok" });
});

export default app;
